src/random_ray/iteration.cpp

```diff
@@ -0,0 +1,660 @@
+      MPI_Send(random_ray::position.data(), random_ray::n_source_regions * 3,
+        MPI_DOUBLE, 0, 0, mpi::intracomm);
+    }
+  }
+
+  // For the rest of the source region data, we simply perform an all reduce,
+  // as these values will be needed on all ranks for transport during the
+  // next iteration.
+  MPI_Allreduce(MPI_IN_PLACE, random_ray::volume.data(),
+    random_ray::n_source_regions, MPI_DOUBLE, MPI_SUM, mpi::intracomm);
+
+  MPI_Allreduce(MPI_IN_PLACE, random_ray::was_hit.data(),
+    random_ray::n_source_regions, MPI_INT, MPI_SUM, mpi::intracomm);
+
+  MPI_Allreduce(MPI_IN_PLACE, random_ray::scalar_flux_new.data(),
+    random_ray::n_source_elements, MPI_FLOAT, MPI_SUM, mpi::intracomm);
+
+  simulation::time_bank_sendrecv.stop();
+#endif
+}
+
+int openmc_run_random_ray()
+{
+  // Initialize OpenMC general data structures
+  openmc_simulation_init();
+
+  // Validate that inputs meet requirements for random ray mode
+  if (mpi::master)
+    validate_random_ray_inputs();
+
+  // Intialize Cell (Flat Source Region) data structures
+  initialize_source_regions();
+
+  // Determine which source term should be used for ray sampling
+  int sampling_source = get_random_ray_sampling_source_index();
+
+  // Random ray mode does not have an inner loop over generations within a
+  // batch, so set the current gen to 1
+  simulation::current_gen = 1;
+
+  // There are no source sites in random ray mode, so be sure to disable to
+  // ensure we don't attempt to write source sites to statepoint
+  settings::source_write = false;
+
+  // If all source regions have been hit by a ray during transport, then we can
+  // skip the mapping process for future iterations.
+  bool mapped_all_tallies = false;
+
+  // Random ray eigenvalue
+  double k_eff = 1.0;
+
+  // Tracks the average FSR miss rate for analysis and reporting
+  double avg_miss_rate = 0.0;
+
+  // Tracks the total number of geometric intersections by all rays for
+  // reporting
+  uint64_t total_geometric_intersections = 0;
+
+  // Begin main simulation timer
+  simulation::time_total.start();
+
+  // Random ray power iteration loop
+  while (simulation::current_batch < settings::n_batches) {
+
+    // Initialize the current batch
+    initialize_batch();
+    initialize_generation();
+
+    // Reset total starting particle weight used for normalizing tallies
+    simulation::total_weight = 1.0;
+
+    // Update source term (scattering + fission)
+    update_neutron_source(k_eff);
+
+    // Reset scalar fluxes, iteration volume tallies, and region hit flags to
+    // zero
+    parallel_fill<float>(random_ray::scalar_flux_new, 0.0f);
+    parallel_fill<double>(random_ray::volume, 0.0);
+    parallel_fill<int>(random_ray::was_hit, 0);
+
+    // Start timer for transport
+    simulation::time_transport.start();
+
+// Transport sweep over all random rays for the iteration
+#pragma omp parallel for schedule(runtime)                                     \
```

Comment:
Is there any scenario where dynamic scheduling would be the wrong choice? Seems like we should not let this be set in runtime to prevent accidental poor scaling with the wrong environment variables.